ledger-tool/src/main.rs

```diff
@@ -2350,6 +2482,72 @@ fn main() {
                     }
                 }
+                ("simulate-block-production", Some(arg_matches)) => {
+                    let marker_file = simulate_block_production_marker_path(&ledger_path);
+                    if !marker_file.exists() {
+                        eprintln!(
+                            "simulate-block-production is for development purposes only. \
+                             It's thus a pretty destructive operation on \
+                             the ledger ({ledger_path:?}). \
+                             Create an empty file at {marker_file:?} if this is intentional."
+                        );
+                        exit(1);
+                    }
+
+                    let mut process_options = parse_process_options(&ledger_path, arg_matches);
+
+                    let banking_trace_events = load_banking_trace_events_or_exit(&ledger_path);
+                    process_options.hash_overrides =
+                        Some(banking_trace_events.hash_overrides().clone());
+
+                    let slot = value_t!(arg_matches, "first_simulated_slot", Slot).unwrap();
+                    let simulator = BankingSimulator::new(banking_trace_events, slot);
+                    let Some(parent_slot) = simulator.parent_slot() else {
+                        eprintln!(
+                            "Couldn't determine parent_slot of first_simulated_slot: {slot} \
+                             due to missing banking_trace_event data."
+                        );
+                        exit(1);
+                    };
+                    process_options.halt_at_slot = Some(parent_slot);
+
+                    let blockstore = Arc::new(open_blockstore(
+                        &ledger_path,
+                        arg_matches,
+                        AccessType::Primary, // needed for purging already existing simulated block shreds...
+                    ));
+                    let genesis_config = open_genesis_config_by(&ledger_path, arg_matches);
+                    let LoadAndProcessLedgerOutput { bank_forks, .. } =
+                        load_and_process_ledger_or_exit(
+                            arg_matches,
+                            &genesis_config,
+                            blockstore.clone(),
+                            process_options,
+                            None, // transaction status sender
+                        );
+
+                    let block_production_method = value_t!(
+                        arg_matches,
+                        "block_production_method",
+                        BlockProductionMethod
+                    )
+                    .unwrap_or_default();
+
+                    info!("Using: block-production-method: {block_production_method}");
+
+                    match simulator.start(
+                        genesis_config,
+                        bank_forks,
+                        blockstore,
```

Comment:
An alternative to make this command less destructive "accidentally" would be to create a new blockstore at some path, maybe `LEDGER_DIR/ledger_tool/simulated_blocks`, and put the newly created blocks in there. This would make replaying them more annoying as some manual intervention would be needed to get those simulated blocks into same Bstore as "original" blocks. One solution to this could be allowing a custom blockstore path under `LEDGER_DIR`, it is currently hard-coded as `LEDGER_DIR/rocksdb`

Reply:
> However, I'm little concerned that we have to ensure banking stage and broadcast stage can work correctly even if separate blockstoresare passed.

I don't think we'd need to pass in two blockstores. Once `load_and_process_ledger_or_exit` returns, we're done with slots older than `bank_forks.read().working_bank().slot()`, which should equal the value we set for `halt_at_slot`. At that point, an empty blockstore is similar to a blockstore that has blocks `[A, Z]` when we're looking to produce blocks for slots `> Z`

Reply:
hmm, that could work. but there's no guarantee for some functionality to start to peek parent slot's shreds for some reason now and in the future? I want to make the subcommand to run under real settings as much as possible to observe approximated runtime behavior.

> This would make replaying them more annoying as some manual intervention would be needed to get those simulated blocks into same Bstore as "original" blocks.

Also, how about this concern?

Overall, i think the current approach is safer and easier..

Reply:
> but there's no guarantee for some functionality to start to peek parent slot's shreds for some reason now and in the future?

A live validator might for stuff like repairs and misc. services, but replay shouldn't. In any case, fair point about something in the future looking further back in ledger.

> > This would make replaying them more annoying as some manual intervention would be needed to get those simulated blocks into same Bstore as "original" blocks.
> >
> Also, how about this concern?

Concern remains / I don't have an answer 😆 

> Overall, i think the current approach is safer and easier.

Yeah, I'm good with the current approach; the idea was offered as brainstorm